### ISSUE TITLE
ActiveRecord::Calculations#pluck does not typecast primary key in `.includes(:association)` when using postgres

### ISSUE THREAD

**Issue by rjclaasen** [rails/rails#43609]
Thank you all for your work on this project & even more gratitude for taking the time to read these words!

In short, chaining a `.pluck` and an `.includes` together can create a query that raises an exception in PostgreSQL. This occurs when the association uses a primary key of type string, while the ID of the association's target record is an int. I would expect the ID to be cast to a string, but this does not occur, causing an exception when the query is performed.

### Steps to reproduce

- Set up a postgres database
- Create two models in the database
- Create an association between them with a primary key of type `string`
- Join the table using `.includes` and use `.pluck` (with any parameter)
- See exception raised by postgres

<details>
<summary>Example executable test case</summary>

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails", branch: "main"
  gem "pg"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This is Postgres specific so we need to create the database if it doesn't exist
# uses the psql command line utilities
DATABASE_NAME = 'bug_test'
if `psql -l | grep #{DATABASE_NAME} | wc -l`.strip.to_i.zero?
  `createdb #{DATABASE_NAME}`
end
ActiveRecord::Base.establish_connection(adapter: "postgresql", database: DATABASE_NAME)
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
  end

  create_table :comments, force: true do |t|
    t.string :post_id
  end
end

class Post < ActiveRecord::Base
  has_many :comments
end

class Comment < ActiveRecord::Base
  belongs_to :post
end

class BugTest < Minitest::Test
  def test_association_stuff
    post = Post.create!
    post.comments << Comment.create!

    assert_equal [1], Post.includes(:comments).pluck(:id)
    # => SELECT "posts"."id" FROM "posts" LEFT OUTER JOIN "comments" ON "comments"."post_id" = "posts"."id"
    #  ActiveRecord::StatementInvalid: PG::UndefinedFunction: ERROR:  operator does not exist: character varying = bigint
    #  LINE 1: ...EFT OUTER JOIN "comments" ON "comments"."post_id" = "posts"....
  end
end
```

</details>

### Expected behavior
ActiveRecord should typecast one of the two IDs so they have identical types in the resulting query. The assertion in the test case is expected to succeed.

### Actual behavior
The resulting query does not include typecasting and an exception is raised.

### System configuration
**Rails version**: 6.1.4.1

**Ruby version**: 3.0.1

**`pg`** version: 1.2.3

**PostgreSQL** version: 14.0

### Further information

I have tested this with a database-independent test case, and it works correctly there. I assume that sqlite3 performs implicit typecasting.

Also occurs when using `.ids` instead of `.pluck`

I'm not sure whether Rails' ActiveRecord or the `pg` gem should be responsible for ensuring this typecasting occurs, so please let me know if this issue should be moved to the `pg` gem!

Thanks again for all your work, please let me know if I can help in any way, I'd love to assist!

**Comment by ghiculescu** [comment]
Hmm, does `Post.includes(:comments).to_a` work (ie. is the issue specific to `pluck`)?

I'm not sure if Active Record explicitly supports mismatched foreign key types. As opposed to SQLite implicitly typecasting.

Out of interest, why is the FK is a string in one table, but the corresponding PK is an integer in the other? What is the use case?

**Comment by rjclaasen** [comment]
`Post.incluedes(:comments).to_a` seems to work, but results in quite a different query compared to `pluck`:

```
=> Post.includes(:comments).to_a
SELECT "posts".* FROM "posts"
   SELECT "comments".* FROM "comments" WHERE "comments"."post_id" = $1  [["post_id", "1"]]

=> Post.includes(:comments).pluck(:id)
SELECT "posts"."id" FROM "posts" LEFT OUTER JOIN "comments" ON "comments"."post_id" = "posts"."id"
```


As for why the primary key needs to be a string, Comments in this case need to support objects with both integer IDs and UUIDs. In practice, it's ActionText records that need to be linked to (parts of) posts or comments. Posts use ID, comments use UUID.

As to why we're using different forms of identification, there's legacy code and half a transition to UUID. 🤷

**Comment by ghiculescu** [comment]
Okay, so it looks like with `Post.includes(:comments).to_a` the type casting is probably being done by the `pg` gem. I'm speculating here but it gets given a string and probably knows how to turn that into an integer before passing it to postgres. Whereas that doesn't work if done via a `JOIN` SQL statement.

If it were me, I'd be doing `.map(&:id)` at this point... I'm not confident this is something that can easily be fixed in Rails.

**Comment by rjclaasen** [comment]
Yeah, it does feel very much like a database-specific edge case.

I probably don't have any more time to look into it further, but my next step would be to see where the connection between Rails and the `pg` gem is made so I can figure out whether the piece of code that would need to be introduced or modified lies within ActiveRecord or the `pg` gem.

**Comment by guaguasi** [comment]
This might be a `pg` issue and is probably fixed in `pg-1.3.0`.

I ran into a problem just now using a UUID as a primary key for a model. `link_to` was translating the uuid to an integer, e.g. `link_to "bar", foo` was generating `/foos/1` instead of `/foos/1ec7186f-335c-61c0-a804-9de580501a5e`. In fact, `web-console` was also reading it that way. In the web console, `Foo.last` was showing `id: 1`, but in `rails c` it was showing as `id: 1ec7186f-335c-61c0-a804-9de580501a5e`. After updating to `pg-1.3.0.rc2` everything worked.
